core-services/src/main/java/org/zalando/nakadi/service/publishing/AvroEventPublisher.java

```diff
@@ -0,0 +1,155 @@
+package org.zalando.nakadi.service.publishing;
+
+import io.opentracing.Span;
+import io.opentracing.tag.Tags;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.EncoderFactory;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.zalando.nakadi.cache.EventTypeCache;
+import org.zalando.nakadi.config.NakadiSettings;
+import org.zalando.nakadi.domain.Envelop;
+import org.zalando.nakadi.domain.EventType;
+import org.zalando.nakadi.domain.NakadiRecord;
+import org.zalando.nakadi.domain.Timeline;
+import org.zalando.nakadi.exceptions.runtime.EventPublishingException;
+import org.zalando.nakadi.exceptions.runtime.InternalNakadiException;
+import org.zalando.nakadi.service.AvroSchema;
+import org.zalando.nakadi.service.TracingService;
+import org.zalando.nakadi.service.timeline.TimelineService;
+import org.zalando.nakadi.service.timeline.TimelineSync;
+import org.zalando.nakadi.util.FlowIdUtils;
+import org.zalando.nakadi.util.UUIDGenerator;
+
+import java.io.ByteArrayOutputStream;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.TimeoutException;
+
+@Service
+public class AvroEventPublisher {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AvroEventPublisher.class);
+
+    private final TimelineService timelineService;
+    private final EventTypeCache eventTypeCache;
+    private final TimelineSync timelineSync;
+    private final NakadiSettings nakadiSettings;
+    private final UUIDGenerator uuidGenerator;
+    private final AvroSchema avroSchema;
+
+    @Autowired
+    public AvroEventPublisher(
+            final TimelineService timelineService,
+            final EventTypeCache eventTypeCache,
+            final TimelineSync timelineSync,
+            final NakadiSettings nakadiSettings,
+            final UUIDGenerator uuidGenerator,
+            final AvroSchema avroSchema) {
+        this.timelineService = timelineService;
+        this.eventTypeCache = eventTypeCache;
+        this.timelineSync = timelineSync;
+        this.nakadiSettings = nakadiSettings;
+        this.uuidGenerator = uuidGenerator;
+        this.avroSchema = avroSchema;
+    }
+
+    public void publishAvro(final String etName,
+                            final String publishedBy,
+                            final GenericRecord event) {
+        try {
+            final GenericRecord metadata = new GenericData.Record(avroSchema.getMetadataSchema());
+            metadata.put("occurred_at", new DateTime(DateTimeZone.UTC).toString());
```

Comment:
Or was the `string` type here intentional?  I would expect `occurred_at` and `received_at` to have the same type.
Here, by the way, we could use the same timestamp instance: there is no distinct "received" moment.